src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestElement.cs

```diff
@@ -109,20 +109,29 @@ public UnitTestElement(TestMethod testMethod)
         /// <returns> An instance of <see cref="TestCase"/>. </returns>
         internal TestCase ToTestCase()
         {
-            var fullName = string.Format(
-                CultureInfo.InvariantCulture,
-                "{0}.{1}",
-                this.TestMethod.FullClassName,
-                this.TestMethod.Name);
+            string fullName = this.TestMethod.HasManagedMethodAndType
+                            ? string.Format(CultureInfo.InvariantCulture, "{0}.{1}", this.TestMethod.ManagedTypeName, this.TestMethod.ManagedMethodName)
+                            : string.Format(CultureInfo.InvariantCulture, "{0}.{1}", this.TestMethod.FullClassName, this.TestMethod.Name);
 
             TestCase testCase = new TestCase(fullName, TestAdapter.Constants.ExecutorUri, this.TestMethod.AssemblyName);
+            if (string.IsNullOrWhiteSpace(this.DisplayName))
+            {
+                testCase.DisplayName = string.IsNullOrWhiteSpace(this.TestMethod.ManagedMethodName) ? this.TestMethod.Name : this.TestMethod.ManagedMethodName;
+            }
+            else
+            {
+                testCase.DisplayName = this.DisplayName;
+            }
```

Comment:
nit: Would be nice to put this logic in a GetDisplayName() function.

Reply:
When would DisplayName be null btw?

Reply:
It's `null` except explicitly set with an attribute.